From 04c646c4c32a77576b10fba214aebcf27081ae9e Mon Sep 17 00:00:00 2001
From: Qiang Yu <yuq825@gmail.com>
Date: Thu, 28 Jun 2018 10:17:39 +0800
Subject: [PATCH 015/151] gallium: add PIPE_CAP_BUFFER_AGE

For gallium drivers to expose EGL/GLX_EXT_buffer_age.

Signed-off-by: Qiang Yu <yuq825@gmail.com>
---
 src/gallium/auxiliary/util/u_screen.c               | 3 +++
 src/gallium/docs/source/screen.rst                  | 1 +
 src/gallium/drivers/swr/swr_screen.cpp              | 1 +
 src/gallium/include/pipe/p_defines.h                | 1 +
 src/gallium/state_trackers/dri/dri_query_renderer.c | 4 +++-
 5 files changed, 9 insertions(+), 1 deletion(-)

diff --git a/src/gallium/auxiliary/util/u_screen.c b/src/gallium/auxiliary/util/u_screen.c
index 399b317..8045d74 100644
--- a/src/gallium/auxiliary/util/u_screen.c
+++ b/src/gallium/auxiliary/util/u_screen.c
@@ -329,6 +329,9 @@ u_pipe_screen_get_param_defaults(struct pipe_screen *pscreen,
    case PIPE_CAP_FORCE_COMPUTE_MINMAX_INDICES:
       return 0;
 
+   case PIPE_CAP_BUFFER_AGE:
+      return 1;
+
    default:
       unreachable("bad PIPE_CAP_*");
    }
diff --git a/src/gallium/docs/source/screen.rst b/src/gallium/docs/source/screen.rst
index 0abd164..91b1c16 100644
--- a/src/gallium/docs/source/screen.rst
+++ b/src/gallium/docs/source/screen.rst
@@ -477,6 +477,7 @@ subpixel precision bias in bits during conservative rasterization.
   0 means no limit.
 * ``PIPE_CAP_MAX_VERTEX_ELEMENT_SRC_OFFSET``: The maximum supported value for
   of pipe_vertex_element::src_offset.
+* ``PIPE_CAP_BUFFER_AGE``: True if the driver wants to expose EGL/GLX_EXT_buffer_age.
 
 .. _pipe_capf:
 
diff --git a/src/gallium/drivers/swr/swr_screen.cpp b/src/gallium/drivers/swr/swr_screen.cpp
index de9008d..eb3264d 100644
--- a/src/gallium/drivers/swr/swr_screen.cpp
+++ b/src/gallium/drivers/swr/swr_screen.cpp
@@ -253,6 +253,7 @@ swr_get_param(struct pipe_screen *screen, enum pipe_cap param)
    case PIPE_CAP_CULL_DISTANCE:
    case PIPE_CAP_CUBE_MAP_ARRAY:
    case PIPE_CAP_DOUBLES:
+   case PIPE_CAP_BUFFER_AGE:
       return 1;
 
    /* MSAA support
diff --git a/src/gallium/include/pipe/p_defines.h b/src/gallium/include/pipe/p_defines.h
index 4d10d1c..974aea41 100644
--- a/src/gallium/include/pipe/p_defines.h
+++ b/src/gallium/include/pipe/p_defines.h
@@ -824,6 +824,7 @@ enum pipe_cap
    PIPE_CAP_MAX_TEXTURE_UPLOAD_MEMORY_BUDGET,
    PIPE_CAP_MAX_VERTEX_ELEMENT_SRC_OFFSET,
    PIPE_CAP_FORCE_COMPUTE_MINMAX_INDICES,
+   PIPE_CAP_BUFFER_AGE,
 };
 
 /**
diff --git a/src/gallium/state_trackers/dri/dri_query_renderer.c b/src/gallium/state_trackers/dri/dri_query_renderer.c
index c7bd428..e86ff5b 100644
--- a/src/gallium/state_trackers/dri/dri_query_renderer.c
+++ b/src/gallium/state_trackers/dri/dri_query_renderer.c
@@ -63,7 +63,9 @@ dri2_query_renderer_integer(__DRIscreen *_screen, int param,
          return -1;
       return 0;
    case __DRI2_RENDERER_EXPOSE_BUFFER_AGE:
-      value[0] = 1;
+      value[0] =
+         screen->base.screen->get_param(screen->base.screen,
+                                        PIPE_CAP_BUFFER_AGE);
       return 0;
    default:
       return driQueryRendererIntegerCommon(_screen, param, value);
-- 
2.7.4

